corehq/util/test_utils.py

```diff
@@ -364,12 +364,20 @@ def unregistered_django_model(model_class):
     return model_class
 
 
-def generate_cases(argsets, cls=None):
-    """Make a decorator to generate a set of parameterized test cases
+class generate_cases:
+    """A decorator to generate parameterized test cases
 
-    Until we have nose generator tests...
+    Usage as test method decorator:
 
-    Usage:
+        class TestThing(TestCase):
+            @generate_cases([
+                ("foo", "bar"),
+                ("bar", "foo"),
+            ])
+            def test_foo(self, foo, bar)
+                self.assertEqual(self.thing[foo], bar)
+
+    Deprecated: two-argument decorator on module level test function:
```

Comment:
I'm curious why the functional usage pattern is deprecated. Is it because nose test generators offer an alternative?

Reply:
The new method decorator pattern replaces both of the old usage patterns, and is more intuitive and in the case of implied test case, more explicit. The new way just seems better in every way. I could have dramatically simplified the implementation (including removing the metaclass magic) if I had replaced all current uses of `generate_cases` with the new pattern, but it's used in a lot of places and I didn't feel like creating that much churn in this PR. Maybe I'll do that in a follow-up?